### PR TITLE
Add SuspendFunWithCoroutineScopeReceiver Rule

### PR DESCRIPTION
**Pull request by hfhbd**
Fixes https://github.com/detekt/detekt/issues/4587

### CODE REVIEW

File: detekt-rules-coroutines/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/coroutines/SuspendFunWithCoroutineScopeReceiverSpec.kt [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+                    launch {
+                        delay(1.seconds)
+                    }
+                }
+                """
+            val findings = subject.compileAndLintWithContext(env, code)
+            assertThat(findings).hasSize(1)
+            assertThat(findings.single().issue.id).isEqualTo(SuspendFunWithCoroutineScopeReceiver.issueID)
+        }
+    }
```
Comment (BraisGabin) [inline detekt-rules-coroutines/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/coroutines/SuspendFunWithCoroutineScopeReceiverSpec.kt:93]: Could you add three more tests to prove that this doesn't flag any function?

- A no-suspend function that have `CoroutineScope` as receiver.
- A suspend function without receiver.
- A suspend function with a `Int` (or any other type) as receiver.

To ensure that we don't have false positives.

**Review by BraisGabin — COMMENTED** [review]
Really nice PR! :)

To make CI happy you need to run `./gradlew generateDocumentation` and commit the changes that it will add in the `default-detekt-config` file.

File: detekt-rules-coroutines/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/coroutines/SuspendFunWithCoroutineScopeReceiverSpec.kt (hunk 1 above)
Comment (hfhbd) [reply detekt-rules-coroutines/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/coroutines/SuspendFunWithCoroutineScopeReceiverSpec.kt:136]: Sure, added

**Comment by hfhbd** [comment]
Ah, there is the default configuration. I've already searched for it :D

**Review by BraisGabin — APPROVED** [review]
LGTM. Thanks for the contribution!

**Comment by hfhbd** [comment]
I forgot the embedded Kotlin compiler provided by Gradle uses a different language API than the actual compiler executing the tests. So you can't use `Duration`, a inline class with `compile-test-snippets=true`.

File: detekt-rules-coroutines/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/coroutines/SuspendFunWithCoroutineScopeReceiverSpec.kt [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+                suspend fun CoroutineScope.foo() {
+                    launch {
+                        delay(timeMillis = 1000)
+                    }
+                }
+                """
+            val findings = subject.compileAndLintWithContext(env, code)
+            assertThat(findings).hasSize(1)
+            assertThat(findings.single().issue.id).isEqualTo(SuspendFunWithCoroutineScopeReceiver.issueID)
```
Comment (cortinico) [inline detekt-rules-coroutines/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/coroutines/SuspendFunWithCoroutineScopeReceiverSpec.kt:34]: small nit: those lines are somehow useless as you are declaring a

```
val subject = SuspendFunWithCoroutineScopeReceiver(Config.empty)
```

just above. Therefore the only issue that can be reported are from this rule

Comment (hfhbd) [reply detekt-rules-coroutines/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/coroutines/SuspendFunWithCoroutineScopeReceiverSpec.kt]: Oh, that's right :D

**Comment by BraisGabin** [comment]
I just found this on the project I work: `fun foo(block: suspend TestCoroutineScope.() -> Unit) {}`. This is basically the same issue. Should we handle it in this rule too?

**Comment by hfhbd** [comment]
@BraisGabin Sure, added

**Comment by BraisGabin** [comment]
Oh, it seems that #4630 is making this PR to fail now. I suppose that they are just the indentation of `"""`.

**Comment by hfhbd** [comment]
Yeah, fixed it.

**Comment by hfhbd** [comment]
@BraisGabin What about changing `compile-test-snippets` to true by default, to run this check with `gradlew build` too?

**Review by cortinico — APPROVED** [review]

**Comment by hfhbd** [comment]
@BraisGabin BTW are you sure about flagging lambdas too? In IntelliJ, only the function with the receiver is marked, but I don't know if this is a bug of IntelliJ...
```kt
import kotlinx.coroutines.*
import kotlin.time.Duration.Companion.seconds

suspend fun main() {
    coroutineScope {
        receiver()
    }
    block {
        launch {

        }
    }
}

suspend fun CoroutineScope.receiver(block: suspend CoroutineScope.() -> Unit) { // Ambiguous coroutineContext due to CoroutineScope receiver of suspend function
    delay(1.seconds)
    launch { // Ambiguous coroutineContext due to CoroutineScope receiver of suspend function
        delay(1.seconds)
        block()
    }
}

suspend fun block(block: suspend CoroutineScope.() -> Unit) {
    delay(1.seconds)
    coroutineScope {
        delay(1.seconds)
        block()
    }
} 
```

**Comment by BraisGabin** [comment]
```kotlin
block {
  launch { } // Which scope am I using the one from suspend or the receiver? 
}
```
I think that the issue is the same. But can you open an issue to intellij to confirm that is a missing part on their end?

**Comment by hfhbd** [comment]
Sure. 
`CoroutineScope`: the inner one wins
`coroutineContext`: this is the problem, because the outer one wins.

**Comment by hfhbd** [comment]
Honestly, after reading https://youtrack.jetbrains.com/issue/KTIJ-12776 and https://youtrack.jetbrains.com/issue/KTIJ-13692/Ambiguous-calls-to-coroutineContext-should-be-reported, I still don't know, when we should mark every coroutine builder (suspend lambda with coroutinescope).
https://youtrack.jetbrains.com/issue/KT-52042/Clarifying-suspend-lambda-and-function-with-coroutineScope-as-pa

**Comment by BraisGabin** [comment]
We can wait to see what they say in the new issue. But for me the issue is the same. In the body of those lambdas is confusing what `coroutineContext` mean. And that's the main reason to use forbid it in `fun`s.

**Comment by hfhbd** [comment]
@BraisGabin Got an interesting answer. https://youtrack.jetbrains.com/issue/KT-52042/Clarifying-suspend-lambda-and-function-with-coroutineScope-as-pa#focus=Comments-27-6029021.0-0
So I would revert marking the lambda parameter too and focus on `suspend fun CoroutineScope.foo` only.

**Pull request merged**